pkg/pushsync/pushsync.go

```diff
@@ -219,80 +224,187 @@ func (ps *PushSync) PushChunkToClosest(ctx context.Context, ch swarm.Chunk) (*Re
+				ps.blocklistPeer(peer)
+			}
+			continue
+		}
 
-	w, r := protobuf.NewWriterAndReader(streamer)
-	if err := ps.sendChunkDelivery(ctx, w, ch); err != nil {
-		_ = streamer.Reset()
-		return nil, fmt.Errorf("chunk deliver to peer %s: %w", peer.String(), err)
-	}
+		receiptRTTTimer := time.Now()
+		receipt, err := ps.receiveReceipt(ctx, r)
+		if err != nil {
+			_ = streamer.Reset()
+			lastErr = fmt.Errorf("receive receipt from peer %s: %w", peer.String(), err)
+			ps.logger.Debugf("pushsync-push: %w", lastErr)
+			if errors.Is(err, context.DeadlineExceeded) {
+				ps.blocklistPeer(peer)
+			}
+			continue
+		}
+		ps.metrics.ReceiptRTT.Observe(time.Since(receiptRTTTimer).Seconds())
+
+		// if you manage to get a tag, just increment the respective counter
+		t, err := ps.tagg.Get(ch.TagID())
+		if err == nil && t != nil {
+			err = t.Inc(tags.StateSent)
+			if err != nil {
+				return nil, err
+			}
+		}
 
-	//  if you manage to get a tag, just increment the respective counter
-	t, err := ps.tagg.Get(ch.TagID())
-	if err == nil && t != nil {
-		err = t.Inc(tags.StateSent)
+		// Check if the receipt is valid
+		if !ch.Address().Equal(swarm.NewAddress(receipt.Address)) {
+			ps.metrics.InvalidReceiptReceived.Inc()
+			_ = streamer.Reset()
+			return nil, fmt.Errorf("invalid receipt. peer %s", peer.String())
+		}
+
+		err = ps.accounting.Credit(peer, receiptPrice)
 		if err != nil {
 			return nil, err
 		}
-	}
 
-	receiptRTTTimer := time.Now()
-	receipt, err := ps.receiveReceipt(ctx, r)
-	if err != nil {
-		_ = streamer.Reset()
-		return nil, fmt.Errorf("receive receipt from peer %s: %w", peer.String(), err)
+		rec := &Receipt{
+			Address: swarm.NewAddress(receipt.Address),
+		}
+
+		return rec, nil
 	}
-	ps.metrics.ReceiptRTT.Observe(time.Since(receiptRTTTimer).Seconds())
 
-	// Check if the receipt is valid
-	if !ch.Address().Equal(swarm.NewAddress(receipt.Address)) {
-		ps.metrics.InvalidReceiptReceived.Inc()
-		_ = streamer.Reset()
-		return nil, fmt.Errorf("invalid receipt. peer %s", peer.String())
+	ps.logger.Tracef("pushsync-push: failed to push chunk %s: reached max peers of %v", ch.Address(), maxPeers)
+
+	if lastErr != nil {
+		return nil, lastErr
 	}
 
-	err = ps.accounting.Credit(peer, receiptPrice)
+	return nil, topology.ErrNotFound
+}
+
+// closestPeer returns address of the peer that is closest to the chunk with
+// provided address addr. This function will ignore peers with addresses
+// provided in skipPeers.
+func (ps *PushSync) closestPeer(addr swarm.Address, skipPeers []swarm.Address) (swarm.Address, error) {
+	closest := swarm.Address{}
+	err := ps.peerSuggester.EachPeerRev(func(peer swarm.Address, po uint8) (bool, bool, error) {
+		for _, a := range skipPeers {
+			if a.Equal(peer) {
+				return false, false, nil
+			}
+		}
+		if closest.IsZero() {
+			closest = peer
+			return false, false, nil
+		}
+		dcmp, err := swarm.DistanceCmp(addr.Bytes(), closest.Bytes(), peer.Bytes())
```

Comment:
I am not sure I get this comment.

I have checked (before completing the PR) about how the `ClosestPeer` works. It is expected that this function always goes through ALL peers, and this compare function then decides which is the "closest".

Reply:
That is true. This is needed as `ClosestPeer` iterates on all peers, as @zbiljic described. This part was required in retrieval also, to avoid upstream requests.

Reply:
oh. my. god. apologies then. our technical debt is way bigger than i thought